### PR TITLE
Fixes #4072: refactor signup code

### LINKED ISSUE #4072: After a user signup we should log him in without redirecting to the sign in screen
There is no reason to redirect to the sign in screen, it was just more convenient. With the addition of magic link and smart lock, it's going to make that redirection not trivial, so we should just kill it.

### PR DESCRIPTION
**Pull request by maxme**
Fixes #4072

I first wanted to factorize the code for sign in (`mFetchBlogListCallback` that is used in both `NewUserFragment` and `SignInFragment`) but this would endup with a lot of branching, so I duplicated it.

Also fixed an animation glitch in 8914ced where enter fragment was on top of the exit fragment for a quick interval during the animation.

I extracted smart lock code into a Helper class so we could reuse it somewhere else (Magic link, other sign in screen for jetpack?). Also moved some logic (connection) from the fragment to the activity so it can be reused when creating a new account.

Needs review: @kwonye since you're going to work on the Smart lock integration in Magic links

### CODE REVIEW

**Comment by mzorz** [comment]
Great code @maxime! While I haven’t found really anything in particular to enhance from your code changes, this thing here caught my attention as we’ve been dealing with language diffs just these days:

In `NewUserFragment.java`, [line 267](https://github.com/wordpress-mobile/WordPress-Android/pull/4234/files#diff-e35ae1bd304722305b741d4480a2de57R267) a new `CreateUserAndBlog` object is created

Within that class, there’s a method called `getDeviceLanguage`, which compares the device’s `Locale.getDefault()` against the list of languages provided by a XML resource (`R.xml.wpcom_languages`), as [can be seen here](https://github.com/wordpress-mobile/WordPress-Android/blob/issue/4072-refactor-signup-code/WordPress/src/main/java/org/wordpress/android/ui/accounts/helpers/CreateUserAndBlog.java#L64)


We'd need to use the new `RestClientUtils.patchDeviceLanguageCode()` method here (as for instance hebrew code in the XML file is `he` as opposed to what `Locale.getDefault()` returns), and even maybe move that method from `RestClientUtils` to some maybe new class like `LanguageUtils` (suggestion)?


On other side of things, a quick search for `Locale.getDefault()` throws several places where we would benefit from using `patchDeviceLanguageCode` to make sure we’re using the latest defined language setting for the device (like, if language settings were changed while our app is running), but this surely is food for a different PR

**Comment by mzorz** [comment]
pinged the wrong Maxime 😱 -  edited now @maxme

**Comment by maxme** [comment]
> Within that class, there’s a method called getDeviceLanguage, which compares the device’s Locale.getDefault() against the list of languages provided by a XML resource (R.xml.wpcom_languages), as can be seen here
> 
> We'd need to use the new RestClientUtils.patchDeviceLanguageCode() method here (as for instance hebrew code in the XML file is he as opposed to what Locale.getDefault() returns), and even maybe move that method from RestClientUtils to some maybe new class like LanguageUtils (suggestion)?

Ah good point. This `getDeviceLanguage` method seems overly complicated, I'll have a deeper look at it, I wonder if we're doing things wrong here.

**Comment by kwonye** [comment]
1. Open the app
2. It should auto-fill, which is correct
3. Background the app
4. Foreground the app

`SignInActivity.onConnected()` always requests credentials, which makes the snack bar show every time. While `SignInFragment.onCredentialsRetrieved()` checks for empty text fields, having the snack bar every time the user comes back to the page is very confusing. 

It's especially confusing in this case:
1. Open the app
2. It should auto-fill, which is correct
3. Change the username
4. Background the app
5. Foreground the app

Can we add a check to `SignInFragment.onConnected()` to only request credentials if the signInFragments' fields are empty?

**Comment by kwonye** [comment]
This enables Smart Lock for the magic links sign in flow, but it does nothing other than fill out the username field.

We can override `SignInActivity.onConnected()` in `MagicLinkSignInActivity` and have it show the password field (which can be achieved by calling `MagicLinkSignInFragment.showPasswordField()`)

**Comment by kwonye** [comment]
Also, I found a bug that's my fault: When clicking "Add self-hosted site" in the Magic Link sign in flow, the "Next" button text never changes. 

Can you please add `mSignInButton.setText(R.string.sign_in);` after `MagicLinkSignInFragment` line `192` please? Thanks in advance.

**Comment by maxme** [comment]
> This enables Smart Lock for the magic links sign in flow, but it does nothing other than fill out the username field.

Ah that's funnny, it wasn't intended. This is a side effect of our fragments (magic or not) using the same fragment tag: https://github.com/wordpress-mobile/WordPress-Android/blob/d705ab6eb2146dcc2eb225a261e7c40a10f22d20/WordPress/src/main/java/org/wordpress/android/ui/accounts/SignInFragment.java#L78-L78

**Comment by maxme** [comment]
> Can we add a check to SignInFragment.onConnected() to only request credentials if the signInFragments' fields are empty?

Yes, I'll ask for credential filling only if the fields are empty.

**Comment by maxme** [comment]
@kwonye: Ready for a second review.

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/SignInFragment.java [hunk 1]
```diff
@@ -640,7 +564,11 @@ public void onResponse(JSONObject jsonObject) {
                         finishCurrentActivity(userBlogList);
                         String displayName = JSONUtils.getStringDecoded(jsonObject, "display_name");
                         Uri profilePicture = Uri.parse(JSONUtils.getString(jsonObject, "avatar_URL"));
-                        saveCredentialsInSmartLock(mUsername, mPassword, displayName, profilePicture);
+                        SmartLockHelper smartLockHelper = getSmartLockHelper();
+                        if (smartLockHelper != null) {
```
Comment (kwonye) [inline WordPress/src/main/java/org/wordpress/android/ui/accounts/SignInFragment.java]: If the user uses Magic Links to sign in, `mUsername` and `mPassword` are null, causing a crash.

**Comment by maxme** [comment]
@kwonye fixed the NPE in 2ad468c 

Also:
* fixed the loading progress indicator in magic link d813b4a 
* disabled auto fill when magic link attempt a login

**Comment by kwonye** [comment]
Works great! Thanks for doing this! SignInActivity was starting to get decently complex so removing the Smart Lock logic really helps readability.

:shipit:

**Pull request merged**